zcash_client_backend/src/fees/common.rs

```diff
@@ -385,106 +381,99 @@ where
         )
         .map_err(|fee_error| ChangeError::StrategyError(E::from(fee_error)))?;
 
-    let fee_with_change = max(
-        fee_without_change,
-        cfg.fee_rule
-            .fee_required(
-                cfg.params,
-                target_height,
-                transparent_input_sizes.clone(),
-                transparent_output_sizes.clone(),
-                sapling_input_count,
-                sapling_output_count(target_change_counts.sapling())?,
-                orchard_action_count(target_change_counts.orchard())?,
-            )
-            .map_err(|fee_error| ChangeError::StrategyError(E::from(fee_error)))?,
-    );
-
-    let total_out_plus_fee_without_change =
-        (total_out + fee_without_change).ok_or_else(overflow)?;
-    let total_out_plus_fee_with_change = (total_out + fee_with_change).ok_or_else(overflow)?;
+    let total_out_with_min_fee = (subtotal_out + min_fee).ok_or_else(overflow)?;
 
     #[allow(unused_mut)]
-    let (mut change, fee) = {
-        if transparent && total_in < total_out_plus_fee_without_change {
-            // Case 1 for a tx with all transparent flows.
-            return Err(ChangeError::InsufficientFunds {
-                available: total_in,
-                required: total_out_plus_fee_without_change,
-            });
-        } else if transparent && total_in == total_out_plus_fee_without_change {
-            // Case 2 for a tx with all transparent flows.
-            (vec![], fee_without_change)
-        } else if total_in < total_out_plus_fee_with_change {
-            // Case 3a, or case 1 or 2 with non-transparent flows.
+    let (mut change, fee) = match total_in.cmp(&total_out_with_min_fee) {
+        Ordering::Less => {
+            // Case 1. Insufficient input value exists to pay the minimum fee; there's no way
+            // we can construct the transaction.
             return Err(ChangeError::InsufficientFunds {
                 available: total_in,
-                required: total_out_plus_fee_with_change,
+                required: total_out_with_min_fee,
             });
-        } else {
-            // Case 3b or 3c.
-            let proposed_change =
-                (total_in - total_out_plus_fee_with_change).expect("checked above");
+        }
+        Ordering::Equal => {
+            // Case 2 for a tx with all transparent flows and no change memo
+            // (e.g. the second transaction of a ZIP 320 pair).
+            (vec![], min_fee)
+        }
```

Comment:
The refactoring to use `.cmp` here removed the `fully_transparent` check for case 2. This contradicts the comment on lines 358-366 and also on line 397. I think the behaviour before that refactor was correct.